performance-tests/bench_2/test_controller/ScenarioManager.cpp

```diff
@@ -201,107 +202,254 @@ void ScenarioManager::customize_configs(std::map<std::string, std::string>& work
   }
 }
 
-AllocatedScenario ScenarioManager::allocate_scenario(
-  const ScenarioPrototype& scenario_prototype, Nodes& available_nodes, bool debug_alloc)
+bool ScenarioManager::is_matched(const std::string& str, const std::string& pat) const
+{
+  if (pat.empty() ||
+      (!OpenDDS::DCPS::is_wildcard(pat.c_str()) && pat == str) ||
+      (OpenDDS::DCPS::is_wildcard(pat.c_str()) && ACE::wild_match(str.c_str(), pat.c_str(), true, true))) {
+    return true;
+  }
+  return false;
+}
+
+AllocatedScenario ScenarioManager::allocate_scenario(const ScenarioPrototype& scenario_prototype,
+  const Nodes& available_nodes, bool debug_alloc)
 {
   if (scenario_prototype.nodes.length() == 0 && scenario_prototype.any_node.length() == 0) {
     throw std::runtime_error("Scenario Prototype is empty!");
   }
 
-  // Get Maximum and Exclusive Node Counts
-  unsigned maximum_nodes = 0;
-  unsigned exclusive_nodes = 0;
-  for (unsigned i = 0; i < scenario_prototype.nodes.length(); i++) {
-    if (scenario_prototype.nodes[i].exclusive) {
-      exclusive_nodes += scenario_prototype.nodes[i].count;
+  // Get a set of matched node controllers for each node prototype
+  typedef std::vector<NodeController::NodeId> MatchedNodeControllers;
+  std::vector<MatchedNodeControllers> matched_ncs(scenario_prototype.nodes.length());
+  for (unsigned i = 0; i < scenario_prototype.nodes.length(); ++i) {
+    const NodePrototype& nodeproto = scenario_prototype.nodes[i];
+    for (Nodes::const_iterator it = available_nodes.begin(); it != available_nodes.end(); ++it) {
+      if (is_matched(it->second.name.in(), nodeproto.name_wildcard.in())) {
+        matched_ncs[i].push_back(it->first);
+      }
     }
-    maximum_nodes += scenario_prototype.nodes[i].count;
-  }
-  // If there are enough nodes, each any node worker can get one to itself.
-  for (unsigned protoworker_i = 0; protoworker_i < scenario_prototype.any_node.length(); protoworker_i++) {
-    maximum_nodes += scenario_prototype.any_node[protoworker_i].count;
   }
 
-  // Get Minimum Node Count
-  unsigned minimum_nodes = 0;
-  minimum_nodes = exclusive_nodes;
-  if (maximum_nodes - exclusive_nodes) { // If there are any non-exclusive node configs
-    minimum_nodes++; // Then there just has to be at least one node for them
+  struct NcMetaData {
+    NcMetaData() : max_prob{0.0},
+                   overlap_count{0},
+                   exclusive_occupied{false}
+    {
+    }
+
+    NodeController::NodeId id;
+
+    // Maximum probabilty that it is used for an exclusive node
+    double max_prob;
+
+    // Number of node prototypes that can assign a node to it
+    unsigned overlap_count;
+
+    // Whether it is occupied by an exclusive node
+    bool exclusive_occupied;
+
+    bool operator()(const NcMetaData& a, const NcMetaData& b) const
```

Comment:
Interesting, this puts the lowest probability with lowest overlap count at the top? Am I reading that right? Maybe a comment here to explain the goal of using this as the sort operator (rather than std::less & operator<).

Reply:
That's correct. I'll add a comment for it.